src/nvim/ex_cmds.c

```diff
@@ -4822,8 +4822,8 @@ static void helptags_one(char_u *dir, char_u *ext, char_u *tagfname,
 
   // Find all *.txt files.
   size_t dirlen = STRLEN(dir);
-  STRCPY(NameBuff, dir);
-  STRCAT(NameBuff, "/**/*");
+  STRLCPY(NameBuff, dir, sizeof(NameBuff));
+  STRCAT(NameBuff, "/**/*");  // NOLINT
```

Comment:
The `+` there makes a difference because it is logically right next to `*`, just outside the group. It is a very bad idea to have two greedy operators right next to each other. You are right that greedy inner match is good in case of success. But to detect failure they try all combinations of splitting the string between them before the regexp gives up. It grows ridiculously fast.

I forgot the exact math, but imagine fibonacci like sequence, where each number is sum of all preceding numbers, not just two. For a string of length n it grows as the nth element in this sequence.